### DISCUSSION TITLE
Translating raw SQL to SQLAlchemy; constraints

### DISCUSSION THREAD

**Discussion by baileythegreen** [sqlalchemy/sqlalchemy#7771] [Usage Questions] [answered]
Hi, I work on a Python tool backed by an SQLite database, which uses SQLAlchemy to interact with the database. I am currently adding Alembic support to the tool.

I am trying to use SQLAlchemy's `declarative_base()` syntax to create a table similar to the one Alembic uses to track the database version. I don't seem to be able to produce the same using SQLAlchemy—though running Alembic's SQL command (which is identical to the target schema) works.

How can I get the `CONSTRAINT <name>` portion to appear in the schema?  Does its presence in one schema, but not the other, mean they are not equivalent to each other?

## Target schema
```sql
CREATE TABLE alembic_version (
	version_num VARCHAR(32) NOT NULL,
	CONSTRAINT alembic_version_pkc PRIMARY KEY (version_num)
);
```

## Attempted SQLalchemy code
```python
from sqlalchemy import Column, types, PrimaryKeyConstraint
from sqlalchemy.ext.declarative import declarative_base  # type: ignore

Base = declarative_base() 

class Alembic(Base):
    __tablename__ = "alembic_version"
    __table_args__ = tuple(PrimaryKeyConstraint('alembic_version_pkc', name='version_num'))

    version_num = Column(types.VARCHAR(32), primary_key=True, nullable=False)
```

## Resulting schema
```sql
CREATE TABLE alembic_version (
	version_num VARCHAR(32) NOT NULL, 
	PRIMARY KEY (version_num)
);
```

**Comment by CaselIT** [comment, marked as answer]
hi,

I think the issue here is `tuple(PrimaryKeyConstraint('alembic_version_pkc', name='version_num'))`. this will return a tuple of the iteration over a primary key constriant that I'm guessing will return the columns.

Try using instead
```py
from sqlalchemy import Column, types, PrimaryKeyConstraint
from sqlalchemy.ext.declarative import declarative_base  # type: ignore

Base = declarative_base() 

class Alembic(Base):
    __tablename__ = "alembic_version"
    __table_args__ = tuple([PrimaryKeyConstraint('alembic_version_pkc', name='version_num')])

    version_num = Column(types.VARCHAR(32), nullable=False)
```

**Reply by baileythegreen** [reply]
Hi, I see your point about that line; (I always try to use `tuple()` as thought it were `set()`; never seem to learn...).

However, I tried your suggestion, (and the same, but with the `primary_key=True` from my original code added back in), but both of these give a `KeyError` (see traceback below). I don't know if there's something else I'm missing, as I have not been able to find an actual example of a `PrimaryKeyConstraint` being set in the `declarative_base()` syntax, just this sentence from the docs (https://docs.sqlalchemy.org/en/14/core/constraints.html#setting-up-constraints-when-using-the-declarative-orm-extension).

> To apply table-level constraint objects such as [ForeignKeyConstraint](https://docs.sqlalchemy.org/en/14/core/constraints.html#sqlalchemy.schema.ForeignKeyConstraint) to a table defined using Declarative, use the __table_args__ attribute, described at [Table Configuration](https://docs.sqlalchemy.org/en/14/orm/extensions/declarative/table_config.html#declarative-table-args).

The `KeyError` seems like the issue is the code is trying to access a column before it is created, but I'm not sure how that squares with this example of other constraint types from the docs (https://docs.sqlalchemy.org/en/14/orm/declarative_tables.html#orm-declarative-table-configuration):

```python
class MyClass(Base):
    __tablename__ = 'sometable'
    __table_args__ = (
            ForeignKeyConstraint(['id'], ['remote_table.id']),
            UniqueConstraint('foo'),
            )
```

## Traceback
```python
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "./sqlalchemy/ext/declarative/api.py", line 75, in __init__
    _as_declarative(cls, classname, cls.__dict__)
  File "./sqlalchemy/ext/declarative/base.py", line 131, in _as_declarative
    _MapperConfig.setup_mapping(cls, classname, dict_)
  File "./sqlalchemy/ext/declarative/base.py", line 160, in setup_mapping
    cfg_cls(cls_, classname, dict_)
  File "./sqlalchemy/ext/declarative/base.py", line 190, in __init__
    self._setup_table()
  File "./sqlalchemy/ext/declarative/base.py", line 534, in _setup_table
    cls.__table__ = table = table_cls(
  File "<string>", line 2, in __new__
  File "./sqlalchemy/util/deprecations.py", line 128, in warned
    return fn(*args, **kwargs)
  File "./sqlalchemy/sql/schema.py", line 506, in __new__
    metadata._remove_table(name, schema)
  File "./sqlalchemy/util/langhelpers.py", line 68, in __exit__
    compat.reraise(exc_type, exc_value, exc_tb)
  File "./sqlalchemy/util/compat.py", line 153, in reraise
    raise value
  File "./sqlalchemy/sql/schema.py", line 501, in __new__
    table._init(name, metadata, *args, **kw)
  File "./sqlalchemy/sql/schema.py", line 600, in _init
    self._init_items(*args)
  File "./sqlalchemy/sql/schema.py", line 117, in _init_items
    spwd(self)
  File "./sqlalchemy/sql/base.py", line 456, in _set_parent_with_dispatch
    self._set_parent(parent)
  File "./sqlalchemy/sql/schema.py", line 3391, in _set_parent
    super(PrimaryKeyConstraint, self)._set_parent(table)
  File "./sqlalchemy/sql/schema.py", line 2891, in _set_parent
    ColumnCollectionMixin._set_parent(self, table)
  File "./sqlalchemy/sql/schema.py", line 2848, in _set_parent
    for col in self._col_expressions(table):
  File "./sqlalchemy/sql/schema.py", line 2842, in _col_expressions
    return [
  File "./sqlalchemy/sql/schema.py", line 2843, in <listcomp>
    table.c[col] if isinstance(col, util.string_types) else col
  File "./sqlalchemy/util/_collections.py", line 194, in __getitem__
    return self._data[key]
KeyError: 'alembic_version_pkc'
```

**Reply by CaselIT** [reply]
sorry it's late.. the correct invocation for `PrimaryKeyConstraint` is `PrimaryKeyConstraint('version_num', name='alembic_version_pkc')`

**Reply by baileythegreen** [reply]
Ah, that's done it! I think I had tried that order before, but prior to fixing the `tuple()` call, so it didn't work.

With both of those changes, I am now getting the correct schema. Thank you!
